tokio/src/task/scope/cancellation_token.rs

```diff
@@ -0,0 +1,255 @@
+//! An asynchronously awaitable event for signalization between tasks
+
+use super::intrusive_double_linked_list::{LinkedList, ListNode};
+use std::{
+    future::Future,
+    pin::Pin,
+    sync::Mutex,
+    task::{Context, Poll, Waker},
+};
+
+/// Tracks how the future had interacted with the event
+#[derive(PartialEq)]
+enum PollState {
+    /// The task has never interacted with the event.
+    New,
+    /// The task was added to the wait queue at the event.
+    Waiting,
+    /// The task has been polled to completion.
+    Done,
+}
+
+/// Tracks the WaitForCancellationFuture waiting state.
+/// Access to this struct is synchronized through the mutex in the Event.
+struct WaitQueueEntry {
+    /// The task handle of the waiting task
+    task: Option<Waker>,
+    /// Current polling state
+    state: PollState,
+}
+
+impl WaitQueueEntry {
+    /// Creates a new WaitQueueEntry
+    fn new() -> WaitQueueEntry {
+        WaitQueueEntry {
+            task: None,
+            state: PollState::New,
+        }
+    }
+}
+
+/// Internal state of the `CancellationToken` pair above
```

Comment:
`CancellationToken` is now below, not above, and is not a pair. Suggest you just make this

~~~
/// Internal state of the `CancellationToken`
~~~